### PR TITLE
[scheduling] Expand slot scheduler to resource scheduler 

### LINKED ISSUE #2787: [Proposal] Expand slot management to resource management
## Motivation

Currently Mars use slot for resource management and bands allocation which just consider cpu/gpu but no memory. Mars always allocate one slot which represents one core cpu or gpu for a subtask. It works well most time. But there are some shortcomings like:

* Subtasks need less cpu but assigned more which results in low cpu utilization and long execution time
* Subtasks need more memory and less cpu which leads node OOM

So we could develop more granular resource management and allocation to increase resource utilization, improve scheduling efficiency, and avoid OOM.

## Design

We propose a more common resource management which includes not only cpu/gpu but also memory, and even estimated time of a subtask.
A subtask of Mars needs one slot but no other resource by default. We could add more different types of resources to management.
Obviusly we can involve memory first as follows:

```
class Resource:
    num_cpus: float
    num_gpus: float
    num_mem_bytes: float
```

With this we can expand slot management to resource management. And bands allocation needs to consider both cpu/gpu and memory.
So we should develop a more complex resource management from a simple resource(cpu/gpu) to multiple resources.

In addition, we can easily implement hbo if we have an external system which can recommend resources for subtasks by history information.
If no external system, we can set memory resource to 0 which degenerates to the original slot scheduler or set a value through configuration to avoid OOM.

And later we can estimated execution time of subtasks if the external HBO system can recommend subtask execution time.

## Plan

In order to implement this proposal, we plan to do:
* Add physical resource management which has been in #2731 
* Add a logic id for subtask which represents a unique subtask and in different submits the same subtask has same logic id which has been in #2575 
* Add a logic id for tileable graph which just like subtask logic id and this is for HBO
* Introduce resource management and bands allocation #2825

### PR DESCRIPTION
**Pull request by zhongchun**
## What do these changes do?

Currently Mars use slot for resource management and bands allocation which just consider cpu/gpu but no memory. Mars always allocate one slot which represents one core cpu or gpu for a subtask. It works well most time. But there are some shortcomings like:

Subtasks need less cpu but assigned more which results in low cpu utilization and long execution time
Subtasks need more memory and less cpu which leads node OOM
So we could develop more granular resource management and allocation to increase resource utilization, improve scheduling efficiency, and avoid OOM.

## Related issue number

Closes #2787 

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/deploy/oscar/worker.py [hunk 1]
```diff
@@ -28,13 +28,16 @@ class WorkerCommandRunner(OscarCommandRunner):
...
-        self.band_to_slot = dict()
+        self.band_to_resource = dict()
         self.cuda_devices = []
         self.n_io_process = 1
 
     def config_args(self, parser):
         super().config_args(parser)
         parser.add_argument("--n-cpu", help="num of CPU to use", default="auto")
+        parser.add_argument(
+            "--n-mem-bytes", help="num in bytes of memory to use", default="auto"
```
Comment (wjsi) [inline mars/deploy/oscar/worker.py:39]:
```suggestion
            "--mem-bytes", help="bytes of memory to use", default="auto"
```

File: mars/deploy/oscar/worker.py [hunk 2]
```diff
@@ -58,6 +61,9 @@ def parse_args(self, parser, argv, environ=None):
         self.n_io_process = int(args.n_io_process)
 
         n_cpu = cpu_count() if args.n_cpu == "auto" else int(args.n_cpu)
+        n_mem_bytes = (
+            mem_total() if args.n_mem_bytes == "auto" else int(args.n_mem_bytes)
+        )
```
Comment (wjsi) [inline mars/deploy/oscar/worker.py:66]:
```suggestion
        mem_bytes = (
            mem_total() if args.mem_bytes == "auto" else int(args.mem_bytes)
        )
```

File: mars/services/task/supervisor/processor.py [hunk 3]
```diff
@@ -24,6 +24,7 @@
 from functools import reduce, wraps
 from typing import Callable, Coroutine, Dict, Iterator, List, Optional, Set, Type, Union
 
+
```
Comment (wjsi) [inline mars/services/task/supervisor/processor.py:27]: Remove redundant empty line

File: mars/services/cluster/uploader.py [hunk 4]
```diff
@@ -154,15 +155,19 @@ async def upload_node_info(self, status: NodeStatus = None):
...
 
     def get_bands(self) -> Dict[BandType, int]:
-        band_slots = dict()
+        band_resource = dict()
         for resource_type, info in self._info.resource.items():
             if resource_type.startswith("numa"):
                 # cpu
-                band_slots[(self.address, resource_type)] = info["cpu_total"]
+                band_resource[(self.address, resource_type)] = Resource(
+                    num_cpus=info["cpu_total"], num_mem_bytes=info["memory_total"]
```
Comment (wjsi) [inline mars/services/cluster/uploader.py:163]:
```suggestion
                    num_cpus=info["cpu_total"], mem_bytes=info["memory_total"]
```

File: mars/services/task/supervisor/processor.py [hunk 5]
```diff
@@ -72,6 +74,25 @@ async def inner(processor: "TaskProcessor", *args, **kwargs):
     return inner
 
 
+def _init_subtask_required_resource(subtask_graph: SubtaskGraph):
```
Comment (wjsi) [inline mars/services/task/supervisor/processor.py:77]: No need to create a separate method.

File: mars/services/subtask/core.py [hunk 6]
```diff
@@ -72,6 +73,7 @@ class Subtask(Serializable):
     logic_parallelism: int = Int32Field("logic_parallelism")
     # subtask can only run in specified bands in `expect_bands`
     bands_specified: bool = BoolField("bands_specified")
+    required_resource: Resource = ReferenceField("required_resource", Resource)
```
[the remark this answers is not in the document]
Comment (zhongchun) [reply mars/services/subtask/core.py:76]: Ok. I'll fix it.

File: mars/services/task/supervisor/processor.py (hunk 3 above)
Comment (zhongchun) [reply mars/services/task/supervisor/processor.py:27]: Thanks for your reminder.

File: mars/services/cluster/uploader.py (hunk 4 above)
Comment (zhongchun) [reply mars/services/cluster/uploader.py:163]: `num_mem_bytes` here is  a `Resource` parameter in `mars/_resource.pyx` which is merged in #2732. How about rename it in a later pr if we are sure to fix it?

File: mars/services/task/supervisor/processor.py (hunk 5 above)
Comment (zhongchun) [reply mars/services/task/supervisor/processor.py:77]: This method is designed to implement different initialization of subtasks' required resource. By default we just initialize  resources with special values, and like HBO, we can get the resources through external service.

File: mars/services/cluster/gather.py [hunk 7]
```diff
@@ -131,29 +133,38 @@ def gather_node_env():
...
     res = dict()
     mem_info = mars_resource.virtual_memory()
     num_cpu = (
         mars_resource.cpu_count()
-        if band_to_slots is None
-        else band_to_slots.get("numa-0", None)
+        if band_to_resource is None
+        else band_to_resource.get("numa-0", ZeroResource).num_cpus
+    )
+    num_mem_bytes = (
```
[the remark this answers is not in the document]
Comment (zhongchun) [reply mars/services/cluster/gather.py]: Fixed. Thanks.

File: mars/deploy/oscar/local.py [hunk 8]
```diff
@@ -51,6 +51,7 @@ async def new_cluster_in_isolation(
     address: str = "0.0.0.0",
     n_worker: int = 1,
     n_cpu: Union[int, str] = "auto",
+    n_mem_bytes: Union[int, str] = "auto",
```
[the remark this answers is not in the document]
Comment (zhongchun) [reply mars/deploy/oscar/local.py]: Fixed. Thanks.

File: mars/deploy/oscar/local.py [hunk 9]
```diff
@@ -65,6 +66,7 @@ async def new_cluster_in_isolation(
         address,
         n_worker,
         n_cpu,
+        n_mem_bytes,
```
[the remark this answers is not in the document]
Comment (zhongchun) [reply mars/deploy/oscar/local.py]: Fixed. Thanks.

File: mars/deploy/oscar/local.py [hunk 10]
```diff
@@ -79,6 +81,7 @@ async def new_cluster(
     address: str = "0.0.0.0",
     n_worker: int = 1,
     n_cpu: Union[int, str] = "auto",
+    n_mem_bytes: Union[int, str] = "auto",
```
[the remark this answers is not in the document]
Comment (zhongchun) [reply mars/deploy/oscar/local.py]: Fixed. Thanks.

File: mars/deploy/oscar/worker.py (hunk 1 above)
Comment (zhongchun) [reply mars/deploy/oscar/worker.py]: Fixed. Thanks.

File: mars/deploy/oscar/worker.py (hunk 2 above)
Comment (zhongchun) [reply mars/deploy/oscar/worker.py]: Fixed. Thanks.

File: .python-version [hunk 11]
```diff
@@ -0,0 +1 @@
+gmars379
```
Comment (wjsi) [inline .python-version:1]: What's this?

Comment (zhongchun) [reply .python-version:1]: Sorry, accidentally added. I'll remove it.

**Review by wjsi — COMMENTED** [review]
Attribute `num_mem_bytes` in `Resource` class shall also be renamed as `mem_bytes`.

File: mars/services/cluster/uploader.py (hunk 4 above)
Comment (wjsi) [reply mars/services/cluster/uploader.py:163]: Just fix it in this PR.

File: mars/services/task/supervisor/processor.py (hunk 5 above)
Comment (wjsi) [reply mars/services/task/supervisor/processor.py:76]: A bare modular method is not friendly for future maintainers. At least it should be a class member.

File: mars/services/cluster/uploader.py (hunk 4 above)
Comment (zhongchun) [reply mars/services/cluster/uploader.py:163]: OK.

File: mars/services/task/supervisor/processor.py (hunk 5 above)
Comment (zhongchun) [reply mars/services/task/supervisor/processor.py:76]: I could add a 'ResourceAssigner' which assigns resources for subtasks by configuration, like default values or external service.

Comment (wjsi) [reply mars/services/task/supervisor/processor.py:76]: Maybe `ResourceEvaluator` could be a better name.

Comment (zhongchun) [reply mars/services/task/supervisor/processor.py:76]: Good suggestion.

**Comment by qinxuye** [comment]
I think failure of asv benchmark can be ignored, because the modification, the benchmark cannot run on master branch any more. The time in #2875 is 1.09±0.02s, I think this PR does not make the time worse.

**Comment by zhongchun** [comment]
> I think failure of asv benchmark can be ignored, because the modification, the benchmark cannot run on master branch any more. The time in #2875 is 1.09±0.02s, I think this PR does not make the time worse.

I think so. Thanks for your reminder.

**Review by wjsi — APPROVED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**